File: import/new/1396.py
# https://leetcode.com/problems/design-underground-system
class UndergroundSystem:
    def __init__(self):
        self.checkIns = {}
        self.times = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.checkIns[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        start_station, start_time = self.checkIns[id]
        if (start_station, stationName) not in self.time:
            self.time[(start_station, stationName)] = [t - start_time, 1]
        else:
            total_time, counts = self.time[(start_station, stationName)]
            self.times[(start_station, stationName)] = [total_time + t - start_time, counts + 1]

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        total_time, counts = self.times[(startStation, endStation)]
        return total_time / counts

File: code/java/1396.java
// https://leetcode.com/problems/design-underground-system
class UndergroundSystem {
    private final Map<Integer, Pair<String, Integer>> checkIns = new HashMap<>();
    private final Map<Pair<String, String>, long[]> times = new HashMap<>();

    public void checkIn(int id, String stationName, int t) {
        checkIns.put(id, new Pair<>(stationName, t));
    }

    public void checkOut(int id, String stationName, int t) {
        Pair<String, Integer> in = checkIns.remove(id);
        String stationStart = in.getKey();
        int dt = t - in.getValue();
        Pair<String, String> key = new Pair<>(stationStart, stationName);
        long[] stats = times.get(key);
        if (stats == null) {
            times.put(key, new long[]{dt, 1});
        } else {
            stats[0] += dt;
            stats[1] += 1;
        }
    }

    public double getAverageTime(String startStation, String endStation) {
        long[] a = times.get(new Pair<>(startStation, endStation));
        return (double) a[0] / a[1];
    }
}
